### DISCUSSION TITLE
Auto-Detect Deno Project Notification

### DISCUSSION THREAD

**Discussion by kitsonk** [denoland/vscode_deno#354] [Ideas]
Wanted to get feedback from the community for a potential feature for the extension/language server.

The idea would be that while opening up a file in a workspace of a non-Deno enabled project, and the Deno language server detects what looks like a potential Deno file, to send a notification to the user that would allow them to click a button to setup the workspace as a Deno enabled workspace.

The detection would be, a JavaScript, TypeScript, JSX or TSX file that:

- Has an import of `http:`, `https:`, `data:` _or_
- Has an import that ends with `.ts`, `.js`, `.tsx`, or `.jsx`.

**Comment by RemyRylan** [comment]
Sorry no one responded to this and it's been so long since you posted it. I think if proper detection is the goal, then both of those methods seem like temporary solutions and not guaranteed.

- Node.js will inevitably eventually support remote imports/exports. It's just a matter of when.
- The loader API in Node.js paves the path to specify the full source path instead of relying on the resolution algorithm... of course TypeScript needs to also allow for using `.ts` and `.tsx` extensions though. I think that's one the TS team will have to eventually concede and allow as a configuration option. I'm sure the loader API within Node is going to prove to be very popular, hopefully that, along with the growth of Deno will get the TS to concede on offering a configuration option.

With that said, I offer two suggestions:

1. Do detection on the use of the `Deno` global. Obviously exclusive to Deno, pretty future-proof.
2. If possible just add a prompt on first-install to allow the user to choose:
  - Enable Deno globally (some developers want to exclusively use Deno)
  - Prompt each time when opening a new workspace/project
  - Manually configure for each project

A bit outside of scope (and I don't know how it would work) but one other possibility is for Deno to somehow offer a complete drop-in replacement for `typescript/lib/tsserver.js`, then in VS Code a user could just set the path to `typescript.tsdk` in VS Code settings to the Deno version and they would be able to chose Deno as their version of TypeScript.

**Reply by kitsonk** [reply]
Things have progressed since opening this.  We introduced `deno.jsonc`/`deno.json`.  This is the clearest indicator that a project is a "Deno" project.  We will likely introduce auto detection of the config file in the CLI and replicate that feature in the extension, which will be the "auto-detection" needed, so largely this proposed feature is redundant and unneeded now.
